### PR TITLE
Use HTTP codes to determine MessagingClientErrorCode

### PR DESCRIPTION
**Pull request by kmcgill88**
https://github.com/invertase/dart_firebase_admin/issues/22

@rrousselGit looking for some feedback on this. If y'all agree on the direction I will add tests.

- I cut down on the amount of `MessagingClientErrorCode`. Felt like some were duplicated.
- Added `int? httpStatusCode` to `MessagingClientErrorCode`.
- Added more docs from Firebase docs.

### CODE REVIEW

**Comment by kmcgill88** [comment]
Could a contributor please provide some feedback on this? I've been using this is a production environment successfully for weeks now.

**Comment by kmcgill88** [comment]
@lesnitsky @Ehesp  I'd love to get feedback on this and come to a conclusion. Thanks!

File: packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart [hunk 1]
```diff
@@ -59,83 +59,173 @@ const _topicMgtServerToClientCode = {
...
+  /// If the error persists, please contact Firebase support.
+  ///
   internal(
     code: 'INTERNAL',
-    'Internal server error.',
+    message: 'Internal server error.',
+    httpStatusCode: 500,
   ),
 
+  ///
```
Comment (rrousselGit) [inline packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart:78]:
```suggestion
```

File: packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart [hunk 2]
```diff
@@ -59,83 +59,172 @@ const _topicMgtServerToClientCode = {
   'UNKNOWN': 'UNKNOWN_ERROR',
 };
 
+///
+/// Enum for handling different Firebase Cloud Messaging error cases per
+/// https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode
+///
 enum MessagingClientErrorCode {
+  ///
```
Comment (rrousselGit) [inline packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart:67]:
```suggestion
```

File: packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart [hunk 3]
```diff
@@ -59,83 +59,172 @@ const _topicMgtServerToClientCode = {
   'UNKNOWN': 'UNKNOWN_ERROR',
 };
 
+///
```
Comment (rrousselGit) [inline packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart:62]:
```suggestion
```

File: packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart [hunk 4]
```diff
@@ -59,83 +59,170 @@ const _topicMgtServerToClientCode = {
   'UNKNOWN': 'UNKNOWN_ERROR',
 };
 
+/// Enum for handling different Firebase Cloud Messaging error cases per
+/// https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode
+///
```
Comment (rrousselGit) [inline packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart:64]:
```suggestion
```

File: packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart [hunk 5]
```diff
@@ -59,83 +59,170 @@ const _topicMgtServerToClientCode = {
...
 enum MessagingClientErrorCode {
+  /// The server encountered an error while trying to process the request.
+  /// You could retry the same request following the requirements listed in UNAVAILABLE
+  /// If the error persists, please contact Firebase support.
+  ///
```
Comment (rrousselGit) [inline packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart:69]:
```suggestion
```

File: packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart [hunk 6]
```diff
@@ -59,83 +59,170 @@ const _topicMgtServerToClientCode = {
...
+  /// Invalid data key: Check that the payload data does not contain a ke
+  /// (such as from, or gcm, or any value prefixed by google) that is used internally
+  /// by FCM. Note that some words (such as collapse_key) are also used by FCM but are
+  /// allowed in the payload, in which case the payload value will be overridden by the FCM value.
+  /// Invalid TTL: Check that the value used in ttl is an integer representing a
+  /// duration in seconds between 0 and 2,419,200 (4 weeks).
+  ///
+  /// Invalid parameters: Check that the provided parameters have the right name and type.
+  ///
```
Comment (rrousselGit) [inline packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart:100]:
```suggestion
```

File: packages/dart_firebase_admin/lib/src/messaging/messaging_api.dart [hunk 7]
```diff
@@ -1050,7 +1050,7 @@ class MessagingPayload {
   MessagingPayload({this.data, this.notification}) {
     if (data == null && notification == null) {
       throw FirebaseMessagingAdminException(
-        MessagingClientErrorCode.invalidPayload,
```
Comment (rrousselGit) [inline packages/dart_firebase_admin/lib/src/messaging/messaging_api.dart:1053]: I don't think any of the changes in this file are right.

File: packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart [hunk 8]
```diff
@@ -59,83 +59,167 @@ const _topicMgtServerToClientCode = {
...
+  /// receive messages.
+  ///
+  /// For all these cases, remove this registration token from the app server and stop using it to send messages.
   unregistered(
     code: 'UNREGISTERED',
-    'App instance was unregistered from FCM. '
-    'This usually means that the token used is no longer valid and a new one must be used.',
-  ),
-
-  authenticationError(
```
Comment (rrousselGit) [inline packages/dart_firebase_admin/lib/src/messaging/fmc_exception.dart:99]: Those codes are official errors which you'd find in the JS implementation too. I don't think removing them makes sense

**Comment by rrousselGit** [comment]
Closed in favour of https://github.com/invertase/dart_firebase_admin/pull/43

Thanks for this though! It helped a lot to understand the problem :)

**Pull request closed without merging**